### PR TITLE
JAMES-3143 CassandraSolveMessageInconsistenciesTask

### PR DESCRIPTION
**Pull request by LanKhuat**
Related: https://github.com/linagora/james-project/pull/3296

### CODE REVIEW

File: server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/routes/SolveMessageInconsistenciesRequestToTask.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+package org.apache.james.webadmin.routes;
+
+import javax.inject.Inject;
+
+import org.apache.james.mailbox.cassandra.mail.task.SolveMessageInconsistenciesService;
+import org.apache.james.mailbox.cassandra.mail.task.SolveMessageInconsistenciesTask;
+import org.apache.james.webadmin.tasks.TaskFromRequestRegistry;
+import org.apache.james.webadmin.tasks.TaskRegistrationKey;
+
+public class SolveMessageInconsistenciesRequestToTask extends TaskFromRequestRegistry.TaskRegistration {
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/routes/SolveMessageInconsistenciesRequestToTask.java:29]: Maybe this class belongs to another commit: `expose message consistency solving task over webadmin`, it should be binded in guice and have a test in webadmin/integration-tests

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/task/SolveMessageInconsistenciesTask.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+    public SolveMessageInconsistenciesTask(SolveMessageInconsistenciesService service) {
+        this.service = service;
+        this.context = new Context();
+    }
+
+    @Override
+    public Result run() {
+        return service.fixMessageInconsistencies(context)
+            .block();
```
Comment (Arsnael) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/task/SolveMessageInconsistenciesTask.java:116]: Should maybe subscribe on the scheduler, like other tasks?

```
return service.fixMessageInconsistencies(context)
            .subscribeOn(Schedulers.elastic())
            .block();
```

File: mailbox/cassandra/src/test/resources/json/solveMessageInconsistencies.additionalInformation.json [hunk 3]
```diff
@@ -0,0 +1,16 @@
+{
+  "type":"solve-message-inconsistencies",
+  "timestamp":"2007-12-03T10:15:30Z",
+  "processedImapUidEntries": 2,
+  "processedMessageIdEntries": 1,
+  "addedMessageIdEntries": 1,
+  "updatedMessageIdEntries": 0,
+  "removedMessageIdEntries": 1,
+  "fixedInconsistencies": [
+    "ComposedMessageId{mailboxId=CassandraId{id=551f0580-82fb-11ea-970e-f9c83d4cf8c2}, messageId=CassandraMessageId{uuid=d2bee791-7e63-11ea-883c-95b84008f979}, uid=MessageUid{uid=1}}",
```
Comment (mbaechler) [inline mailbox/cassandra/src/test/resources/json/solveMessageInconsistencies.additionalInformation.json:10]: this serialization is not ok: we should leverage json format if we need to serialize several information

File: server/container/guice/protocols/webadmin-mailbox/src/main/java/org/apache/james/modules/server/MessagesRoutesModule.java [hunk 4]
```diff
@@ -30,5 +30,7 @@
     protected void configure() {
         Multibinder<Routes> routesMultibinder = Multibinder.newSetBinder(binder(), Routes.class);
         routesMultibinder.addBinding().to(MessagesRoutes.class);
+
+
```
Comment (Arsnael) [inline server/container/guice/protocols/webadmin-mailbox/src/main/java/org/apache/james/modules/server/MessagesRoutesModule.java:34]: What are those extra lines for?

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/MessageRoutesTest.java [hunk 5]
```diff
@@ -83,7 +84,9 @@ void beforeEach() {
                 new MessagesRoutes(taskManager,
                     new InMemoryMessageId.Factory(),
                     new MessageIdReIndexerImpl(reIndexerPerformer),
-                    jsonTransformer))
+                    jsonTransformer,
+                    ImmutableSet.of()
+                    ))
```
Comment (Arsnael) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/MessageRoutesTest.java:89]: You can put those trailing parenthesis above

File: server/protocols/webadmin-integration-test/distributed-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/rabbitmq/RabbitMQWebAdminServerIntegrationTest.java [hunk 6]
```diff
@@ -198,6 +198,39 @@ void solveMailboxInconsistenciesTaskShouldBeExposed() {
...
 
+    @Test
+    void solveMessageInconsistenciesTasksShouldBeExposed() {
+        String taskId = with().post(UPGRADE_TO_LATEST_VERSION)
+            .jsonPath()
+            .get("taskId");
+
+        with()
+            .get("/tasks/" + taskId + "/await")
+            .then()
```
Comment (Arsnael) [inline server/protocols/webadmin-integration-test/distributed-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/rabbitmq/RabbitMQWebAdminServerIntegrationTest.java:209]: usually we indent `when()`, `.then()` etc a step back, to have a better visibility with those different steps

File: server/protocols/webadmin-integration-test/distributed-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/rabbitmq/RabbitMQWebAdminServerIntegrationTest.java [hunk 7]
```diff
@@ -198,6 +198,39 @@ void solveMailboxInconsistenciesTaskShouldBeExposed() {
...
+
+        taskId = with()
+            .queryParam("task", "SolveInconsistencies")
+            .post("/messages")
+            .jsonPath()
+            .get("taskId");
+
+        given()
+            .basePath(TasksRoutes.BASE)
+            .when()
```
Comment (Arsnael) [inline server/protocols/webadmin-integration-test/distributed-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/rabbitmq/RabbitMQWebAdminServerIntegrationTest.java:220]: indent

File: server/protocols/webadmin-integration-test/distributed-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/rabbitmq/RabbitMQWebAdminServerIntegrationTest.java [hunk 8]
```diff
@@ -198,6 +198,39 @@ void solveMailboxInconsistenciesTaskShouldBeExposed() {
...
+            .get(taskId + "/await")
+            .then()
```
Comment (Arsnael) [inline server/protocols/webadmin-integration-test/distributed-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/rabbitmq/RabbitMQWebAdminServerIntegrationTest.java:222]: indent

File: server/protocols/webadmin-integration-test/distributed-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/rabbitmq/RabbitMQWebAdminServerIntegrationTest.java [hunk 9]
```diff
@@ -198,6 +198,39 @@ void solveMailboxInconsistenciesTaskShouldBeExposed() {
             .body("additionalInformation.conflictingEntries", hasSize(0));
     }
 
+    @Test
+    void solveMessageInconsistenciesTasksShouldBeExposed() {
+        String taskId = with().post(UPGRADE_TO_LATEST_VERSION)
+            .jsonPath()
+            .get("taskId");
```
Comment (chibenwa) [inline server/protocols/webadmin-integration-test/distributed-webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/rabbitmq/RabbitMQWebAdminServerIntegrationTest.java:205]: This 'upgrade to latest version' should not be needed

File: server/container/guice/protocols/webadmin-mailbox/src/main/java/org/apache/james/modules/server/MessagesRoutesModule.java (hunk 4 above)
Comment (chibenwa) [inline server/container/guice/protocols/webadmin-mailbox/src/main/java/org/apache/james/modules/server/MessagesRoutesModule.java:34]: You miss this line:

```
Multibinder.newSetBinder(binder(),
                TaskFromRequestRegistry.TaskRegistration.class, Names.named(MessagesRoutes.ALL_MESSAGES_TASKS));
```

Declare an empty multibinder if none provided.

Ill be essential for memory that have /messages but probably no tasks for it...

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/task/MessageInconsistenciesEntry.java [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    public static Builder.RequireMailboxId builder() {
+        return mailboxId -> messageId -> messageUid -> new MessageInconsistenciesEntry(mailboxId, messageId, messageUid);
+    }
+
+    private final String mailboxId;
+    private final String messageId;
+    private final Long messageUid;
+
+    private MessageInconsistenciesEntry(@JsonProperty("mailboxId") String mailboxId, @JsonProperty("messageId") String messageId,
+                                       @JsonProperty("uid") Long messageUid) {
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/task/MessageInconsistenciesEntry.java:54]: Putting one parameter per line would ease readability IMO

**Comment by chibenwa** [comment]
```
[19951a0b1ebde0941fd9a7962ddf4c648518e920] [ERROR] Failures: 
[19951a0b1ebde0941fd9a7962ddf4c648518e920] [ERROR]   RabbitMQWebAdminServerIntegrationTest.solveMessageInconsistenciesTasksShouldBeExposed:210 1 expectation failed.
[19951a0b1ebde0941fd9a7962ddf4c648518e920] JSON path status doesn't match.
[19951a0b1ebde0941fd9a7962ddf4c648518e920] Expected: is "completed"
[19951a0b1ebde0941fd9a7962ddf4c648518e920]   Actual: null
```

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/task/SolveMessageInconsistenciesTask.java (hunk 2 above)
Comment (mbaechler) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/task/SolveMessageInconsistenciesTask.java:116]: Not sure, the caller of `run` is supposed to take care of running the task in a dedicated thread, don't you think?

File: src/site/markdown/server/manage-webadmin.md [hunk 11]
```diff
@@ -697,6 +697,64 @@ The scheduled task will have the following type `messageId-reindexing` and the f
...
 
+### Fixing message inconsistencies
+
+This task is only available on top of Guice Cassandra products.
+
+```
+curl -XPOST /messages?task=SolveInconsistencies
+```
+
+Will schedule a task for fixing message inconsistencies between ImapUid and MessageId table.
```
Comment (mbaechler) [inline src/site/markdown/server/manage-webadmin.md:708]: What does it mean for a user? I mean, our users are not supposed the know James cassandra datamodel

File: src/site/markdown/server/manage-webadmin.md [hunk 12]
```diff
@@ -697,6 +697,64 @@ The scheduled task will have the following type `messageId-reindexing` and the f
...
+      "uid": 3
+    }
+  ]
+}
+```
+
+User actions concurrent to the inconsistency fixing task could result in concurrency issues. New inconsistencies 
+could be created. 
+
+However table of truth (ImapUidTable) would not be impacted hence rerunning the task will eventually fix all issues.
```
Comment (mbaechler) [inline src/site/markdown/server/manage-webadmin.md:753]: could you talk about "source of truth" rather than "table of truth"?

**Comment by chibenwa** [comment]
```
[8c2c42a7e1a82c3cc9394583033cea69e9200a04] Connecting to github.com (github.com)|140.82.118.4|:443... connected.
[8c2c42a7e1a82c3cc9394583033cea69e9200a04] HTTP request sent, awaiting response... 500 internal server error
[8c2c42a7e1a82c3cc9394583033cea69e9200a04] 2020-04-23 14:28:53 ERROR 500: internal server error.
```

test this please

**Comment by chibenwa** [comment]
```
[8c2c42a7e1a82c3cc9394583033cea69e9200a04] + docker logs james-server-cassandra-rabbitmq-cassandra-message-inconsistencies-task-11baa46f-c759-481a-90bb-38ee99f092d8
[8c2c42a7e1a82c3cc9394583033cea69e9200a04] Error: A JNI error has occurred, please check your installation and try again
[8c2c42a7e1a82c3cc9394583033cea69e9200a04] Exception in thread "main" java.lang.UnsupportedClassVersionError: org/apache/james/CassandraRabbitMQJamesServerMain has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0
[8c2c42a7e1a82c3cc9394583033cea69e9200a04] 	at java.lang.ClassLoader.defineClass1(Native Method)
[8c2c42a7e1a82c3cc9394583033cea69e9200a04] 	at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
[8c2c42a7e1a82c3cc9394583033cea69e9200a04] 	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
[8c2c42a7e1a82c3cc9394583033cea69e9200a04] 	at java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
[8c2c42a7e1a82c3cc9394583033cea69e9200a04] 	at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
[8c2c42a7e1a82c3cc9394583033cea69e9200a04] 	at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
[8c2c42a7e1a82c3cc9394583033cea69e9200a04] 	at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
[8c2c42a7e1a82c3cc9394583033cea69e9200a04] 	at java.security.AccessController.doPrivileged(Native Method)
[8c2c42a7e1a82c3cc9394583033cea69e9200a04] 	at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
[8c2c42a7e1a82c3cc9394583033cea69e9200a04] 	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
[8c2c42a7e1a82c3cc9394583033cea69e9200a04] 	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
[8c2c42a7e1a82c3cc9394583033cea69e9200a04] 	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
[8c2c42a7e1a82c3cc9394583033cea69e9200a04] 	at sun.launcher.LauncherHelper.checkAndLoadMain(LauncherHelper.java:495)
```

(Rebase on master needed)

File: src/site/markdown/server/manage-webadmin.md [hunk 13]
```diff
@@ -705,7 +705,24 @@ This task is only available on top of Guice Cassandra products.
 curl -XPOST /messages?task=SolveInconsistencies
 ```
 
-Will schedule a task for fixing message inconsistencies between ImapUid and MessageId table. 
+Will schedule a task for fixing message inconsistencies created by the message denormalization process. 
+
+Messages are denormalized and stored in separated data tables in Cassandra, so they can be accessed 
+by their unique identifier or mailbox identifier & unique identifier (within that mailbox) through different protocols.
```
Comment (mbaechler) [inline src/site/markdown/server/manage-webadmin.md:711]:
```suggestion
by their unique identifier or mailbox identifier & local mailbox identifier through different protocols. 
```

File: src/site/markdown/server/manage-webadmin.md [hunk 14]
```diff
@@ -705,7 +705,24 @@ This task is only available on top of Guice Cassandra products.
...
+```
+BOB receives a message
+The denormalization process fails
+BOB can read the message via JMAP
+BOB cannot read the message via IMAP
+
+BOB marks a message as SEEN
+The denormalization process fails
+The message is SEEN in JMAP
+The message is UNSEEN in IMAP
```
Comment (mbaechler) [inline src/site/markdown/server/manage-webadmin.md:724]:
```suggestion
The message is SEEN via JMAP
The message is UNSEEN via IMAP
```

**Comment by LanKhuat** [comment]
rebased

**Comment by chibenwa** [comment]
Merged
